### PR TITLE
WIP - Update badger to v2

### LINKED ISSUE #2130: Move to badger v2
When debugging a problem with the badger storage that an internal team at Red Hat was having, I changed our code to use Badger v2 and I have it working on a local branch.

One thing that needs to be decided is about the data migration, as the database from v1 isn't compatible with Badger v2. Should we attempt to provide an automatic migration tool? This can potentially take hours and might require an admin intervention to add resources to this one-off process. For instance, my laptop (16GiB RAM) was completely taken up by this for about 40 minutes and failed several times. This would also require backing up the data before starting the process, which probably means provisioning a separate volume in container environments. In the end, users might decide that it's worth just starting fresh with a new database.

During this migration, Jaeger has to be offline, no matter what we decide.

This is basically what needs to be done: https://github.com/dgraph-io/badger/blob/master/README.md#i-see-manifest-has-unsupported-version-x-we-support-y-error

cc @aditya-konarde, @burmanm

### PR DESCRIPTION
**Pull request by jpkrohling**
Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

## Which problem is this PR solving?
- Closes #2130

## Short description of the changes
- [x] Changed the badger dependency to use the v2 module
- [ ] Change the factory to use the new `With*` functions
- [ ] Integrate badger logger with zap
- [ ] Use InMemory instead of tmpfs when in ephemeral mode

### CODE REVIEW

File: plugin/storage/badger/spanstore/cache_test.go [hunk 1]
```diff
@@ -107,10 +107,10 @@ func TestOldReads(t *testing.T) {
 
 // func runFactoryTest(tb testing.TB, test func(tb testing.TB, sw spanstore.Writer, sr spanstore.Reader)) {
 func runWithBadger(t *testing.T, test func(store *badger.DB, t *testing.T)) {
-	opts := badger.DefaultOptions
+	dir, _ := ioutil.TempDir("", "badger")
```
Comment (burmanm) [inline plugin/storage/badger/spanstore/cache_test.go]: With v2, we want to move away from the tempdirs and tmpfs. Instead, what should be used for in-memory situations is: ``badger.DefaultOptions("").WithInMemory(true)`` and this applies to tests also.

**Comment by burmanm** [comment]
Same actually for other opts issues, you need to use ``.WithXXXX()`` methods to configure it instead of those old ``opts.XXXX``

**Comment by objectiser** [comment]
@jpkrohling Regarding the upgrade question - as v2 uses a different package, could we (for a couple of minor releases) have both v1 and v2 inside the badger storage, but v1 just as read only? Possibly with some intelligence to know what date/time the changeover happened - so that only queries that start before that date/time will also be applied to the v1 storage?

File: plugin/storage/badger/spanstore/cache_test.go (hunk 1 above)
Comment (jarifibrahim) [reply plugin/storage/badger/spanstore/cache_test.go:110]: @burmanm @jpkrohling In-memory mode would need more memory. Consider using tmpfs if you have memory constraints.

**Comment by jarifibrahim** [comment]
> Same actually for other opts issues, you need to use `.WithXXXX()` methods to configure it instead of those old `opts.XXXX`
> 
> For example, the `opts.SyncWrites = ..` section turns into:
> 
> `opts := badger.DefaultOptions(f.Options.primary.KeyDirectory).WithValueDir(f.Options.primary.ValueDirectory).WithSyncWrites(f.Options.primary.SyncWrites)`

@burmanm We support both ways of setting options. `opt.xxx = yyy` as well as `options.Withxxx(yyy)` are both valid ways of setting the options.

> @jpkrohling Regarding the upgrade question - as v2 uses a different package, could we (for a couple of minor releases) have both v1 and v2 inside the badger storage, but v1 just as read only? Possibly with some intelligence to know what date/time the changeover happened - so that only queries that start before that date/time will also be applied to the v1 storage?

That's also a way to do it but then you'd have to build a Get API which does the following
```
1. Lookup in Badger v2 DB for the key
2. If v2 lookup failed, perform the lookup in badger v1
```

**Comment by burmanm** [comment]
> @burmanm We support both ways of setting options. opt.xxx = yyy as well as options.Withxxx(yyy) are both valid ways of setting the options.

I assume this was changed at some point after 1.6.0? I think in my 1.6 branch I noticed that the old methods did not work. But either way is fine, it's not a big change in my opinion and it's not visible to users of Jaeger.

File: plugin/storage/badger/spanstore/cache_test.go (hunk 1 above)
Comment (burmanm) [reply plugin/storage/badger/spanstore/cache_test.go:110]: tmpfs is what it currently uses - but to emulate the in-memory mode. Within docker containers I would personally opt for in-memory mode if such is available, given that some docker containers could be configured to map tmpfs to a swap or persistent storage, in which case the performance is a bit unpredictable. Or that they don't have tmpfs at all (that's possible also). 

Or does using in-memory mode prevent using compression or other methods that would reduce the memory usage?

**Comment by jpkrohling** [comment]
I'm closing this PR, as I'm not sure this is still something we want to pursue.

**Pull request closed without merging**